# Introduction to Python

# Lecture 1: Hello world, conditional expressions, loops, and lists

## Learning objectives:

At the end of this lecture, you will:

* Execute Python *statements* from within Google Colab notebook.
* Explain what a *variable* is and express a mathematical formula in code.
* Print program outputs.
* Access mathematical functions from a Python module.
* Write your own *function*.
* Form a *condition* using a *boolean expression*.
* Use a conditional expression in combination with a `while`-loop to perform repetitive tasks.
* Store data elements within a Python `list`.
* Use a `for`-loop to iterate, and perform some task over a sequence of elements.

## Programming a mathematical formula

Here is a formula for the position of a ball, $y(t)$, in vertical motion, starting at ground level (i.e. at $y=0$) at time $t=0$:

$$ y(t) = v_{0}t- \frac{1}{2}gt^2, $$

where:

* $y(t)$ is the height (position) as a function of time $t$,
* $v_0$ is the initial velocity (at $t=0$), and
* $g$ is the acceleration due to gravity.

The computational task we want to solve is: given the values of $v_0$, $g$, and $t$, compute the height $y$.

**How do we program this task?** A program is a sequence of instructions given to the computer. However, while a programming language is much **simpler** than a natural language, it is more **pedantic**. Programs must have correct syntax, i.e. correct use of the computer language grammar rules, and no misprints.

So let us execute a Python statement based on this example to evaluate $y(t) = v_0t- \frac{1}{2}gt^2$ for $v_0 = 5 \,\text{ms}^{-1}$, $g = 9.81 \,\text{ms}^{-2}$ and $t = 0.6 \,\text{s}$. If you were doing this on paper, you would probably write something like this: $$y = 5\cdot 0.6 - \frac{1}{2}\cdot 9.81 \cdot 0.6^2.$$ Happily, writing this in Python is very similar:

In [23]:
# Comment: This is a 'code' cell within Jupyter notebook.
# Press Shift-Enter to execute the code within it,
# or click 'Run' in the Jupyter toolbar above.

print(5*0.6 - 0.5*9.81*0.6**2)

1.2342


You probably noticed that, in the code cell we just wrote, the first few lines start with the hash (`#`) character. In Python, we use `#` to tell the Python interpreter to ignore everything that comes after `#`. We call those lines **comments**, and we write them to help us (humans) understand the code. Besides, we used `print` and we enclosed our calculation within parentheses to display the output. We will explain *comments* and *print function* in more detail later.

## Exercise 1.1 during lecture: Open a code cell and write some code.


In [24]:
print(5*0.6 - 0.5*9.81*0.6**2)

1.2342


## Storing numbers in variables

From mathematics, you are already familiar with *variables* (e.g. $v_0 = 5$, $g = 9.81$, $t = 0.6$, $y = v_0t - \frac{1}{2}gt^2$), and know how important they are for working out complicated problems. Similarly, you can use variables in a program to make it easier to read and understand.

In [25]:
v0 = 5
g = 9.81
t = 0.6
y = v0*t - 0.5*g*t**2
print(y)

1.2342


This program performs the same calculations as the previous one and gives the same output. However, this program spans several lines and uses variables.

We usually use one letter for a variable in mathematics, resorting to using the Greek alphabet and other characters for more clarity. The main reason for this is to avoid becoming exhausted from writing when working out long expressions or derivations. However, when programming, you should use more descriptive names for variables. This might not seem like an important consideration for the trivial example here. Still, it becomes increasingly important as the program gets more complicated and if someone else has to read your code.

### Good variable names make a program easier to understand!

Python *allows* variable names to include:
* lowercase `a-z`, uppercase `A-Z`, underscore (`_`), and digits `0-9`, **but** the name cannot start with a digit or contain a whitespace.
* "textlike" [Unicode](http://www.unicode.org/) characters from other languages are also allowed.

Variable names are case-sensitive (strictly, character sensitive, i.e. `a` is different from `A`).

**Good** variable names are often:
* standard one-letter symbols,
* words or abbreviations of words,
* phrases joined together in [snake_case](https://en.wikipedia.org/wiki/Snake_case) or [camelCase](https://en.wikipedia.org/wiki/Camel_case).

Let us rewrite the previous example using more descriptive variable names:

In [26]:
initial_velocity = 5
acceleration_of_gravity = 9.81
TIME = 0.6
VerticalPositionOfBall = initial_velocity*TIME - 0.5*acceleration_of_gravity*TIME**2
print(VerticalPositionOfBall)

1.2342


In Python, you can check if the name you would like to give to your variable is valid or not, i.e. check if it is a valid identifier. To do that, you can enclose the variable name in quotes (to form a string literal) and call its `isidentifier()` method:

In [27]:
print("initial_velocity".isidentifier())  # snake_case-style variable name
print("InitialVelocity".isidentifier())  # CamelCase-style variable name
print("initial velocity".isidentifier())  # variable name contains space

True
True
False


Certain words have a **special meaning** in Python and **cannot be used as variable names**. We refer to them as **keywords**, and in Python 3.9 they are:

`and`, `as`, `assert`, `async`, `await`, `break`, `class`, `continue`, `def`, `del`, `elif`, `else`, `except`, `finally`, `for`, `from`, `global`, `if`, `import`, `in`, `is`, `lambda`, `nonlocal`, `not`, `or`, `pass`, `raise`, `return`, `try`, `while`, `with`, and `yield`.

Similarly, `True`, `False`, and `None` are keywords we use for the values of variables, and we cannot use them for variable names. Keywords are very important in programming and, in these lectures, we will learn how to use some of them.

Finally, Python has some "builtin" functions which are always available. While Python will let you do it, it is usually a bad idea to use these names, since this would *overshadow* the builtin function and prevent us from calling it. The full list of builtin functions in Python 3.9 is `abs()`, `aiter()`, `all()`, `any()`, `anext()`, `ascii()`, `bin()`, `bool()`, `breakpoint()`, `bytearray()`, `bytes()`, `callable()`, `chr()`, `classmethod()`, `compile()`, `complex()`, `delattr()`, `dict()`, `dir()`, `divmod()`, `enumerate()`, `eval()`, `exec()`, `filter()`, `float()`, `format()`, `frozenset()`, `getattr()`, `globals()`, `hasattr()`, `hash()`, `help()`, `hex()`, `id()`, `input()`, `int()`, `isinstance()`, `issubclass()`, `iter()`, `len()`, `list()`, `locals()`, `map()`, `max()`, `memoryview()`, `min()`, `next()`, `object()`, `oct()`, `open()`, `ord()`, `pow()`, `print()`, `property()`, `range()`, `repr()`, `reversed()`, `round()`, `set()`, `setattr()`, `slice()`, `sorted()`, `staticmethod()`, `str()`, `sum()`, `super()`, `tuple()`, `type()`, `vars()`, and  `zip()`.

Please note that the list of keywords and builtins may differ between different Python versions.

## Adding comments to code

Not everything written in a computer program is intended for execution. In Python, anything on a line after the `#` character is ignored and is known as a **comment**. You can write whatever you want in a comment. Comments are intended to be used to explain what a snippet of code is intended for. It might, for example, explain the objective or provide a reference to the data or algorithm used. This is useful for you when you have to understand your code at some later stage, and indeed for whoever has to read and understand your code later.

In [28]:
# Program for computing the height of a ball in vertical motion.
v0 = 5    # Set initial velocity in m/s.
g = 9.81  # Set acceleration due to gravity in m/s^2.
t = 0.6   # Time at which we want to know the height of the ball in seconds.
y = v0*t - 0.5*g*t**2  # Calculate the vertical position.
print(y)

1.2342


## Formatted printing style

Often we want to print out results using a combination of text and numbers, e.g. `"At t=0.6 s, y is 1.23 m"`. In Python, we can do this using f-strings:

In [29]:
print(f"At t={t} s, y is {y} m.")  # f-string method - string literal begins with an f

At t=0.6 s, y is 1.2342 m.


We enclose our sentence in (single or double) quotes to denote a string literal and add `f` in front of it to tell Python to replace `{t}` and `{y}` with the values of `t` and `y`, respectively.

When printing out floating-point numbers, we should **never** quote numbers to a higher accuracy than they were measured. Python provides a *printf formatting* syntax exactly for this purpose. We can see in the following example where the *format* `g` expresses the floating-point number with the minimum number of significant figures, and the *format* `.2f` specifies that only two digits are printed out after the decimal point. We specify the format inside `{}` and separate it from the variable name with `:`.

In [30]:
print(f"At t={t:g} s, y is {y:.2f} m.")  # f-string with specified formatting

At t=0.6 s, y is 1.23 m.


Sometimes we want a multi-line output. This is achieved using a triple quotation, i.e. `"""`:

In [31]:
print(f"""At t={t:f} s, a ball with
initial velocity v0={v0:.3E} m/s
is located at the height y={y:.2f} m.
""")

At t=0.600000 s, a ball with
initial velocity v0=5.000E+00 m/s
is located at the height y=1.23 m.



Notice in this example we used `f`, `.3E`, and `.2f` to define formats, into which we inserted the values of `t`, `v0`, and `y` respectively. You can find more details about the format specification mini-language in the Python [documentation](https://docs.python.org/3/library/string.html#format-specification-mini-language).

Instead of using the f-string formatted printing, Python offers another two syntax alternatives: string's `format` method and the `%` operator. Let us have a look at how we can print `"At t=0.6 s, y is 1.23 m"` using these two alternative solutions.

In [32]:
print("At t={:g} s, y is {:.2f} m.".format(t, y))  # string's format method
print("At t=%g s, y is %.2f m." % (t, y))  # % operator

At t=0.6 s, y is 1.23 m.
At t=0.6 s, y is 1.23 m.


Notice that we defined slots in a string using curly braces `{}` where we also specified the formatting style in the same way we did before. We inserted the values into the slots by passing them to the `format()` method or by writing them in curly braces.

The `%` operator expands out the input tuple place by place (so the first *slot* gets the first element, the second the second, and so on). If there is only one *slot* then using a tuple is optional.

## How are arithmetic expressions evaluated?
Consider the random mathematical expression:

$$ \frac{5}{9} + \frac{3a^4}{2} $$

implemented in Python as `5/9 + 3 * a**4 / 2`. The rules for evaluating the expression are the same as in mathematics: proceed term by term (additions/subtractions) from the left, compute powers first, then multiplication and division. Therefore in this example the order of evaluation will be:

1. `r1 = 5/9`
2. `r2 = a**4`
3. `r3 = 3*r2`
4. `r4 = r3/2`
5. `r5 = r1 + r4`

We use parenthesis to override these default rules. Indeed, many programmers use parenthesis for greater clarity.

## Standard mathematical functions

What if we need to compute $\sin x$, $\cos x$, $\ln x$, $e^x$ etc., in a program? Such functions are available in Python's `math` module. In fact, there is a vast universe of functionality for Python available in modules. We just `import` in whatever we need for the task at hand.

In this example, we compute $\sqrt{2}$ using the `sqrt` function from the `math` module:

In [33]:
import math

# Since we imported library (import math),
# we access the sqrt function using math.sqrt.
r = math.sqrt(2)
print(r)

1.4142135623730951


or:

In [34]:
from math import sqrt

# This time, we did not import the entire library -
# we imported only sqrt function.
# Therefore, we can use it directly.
r = sqrt(2)
print(r)

1.4142135623730951


Let us now have a look at a more complicated expression, such as

$$\sin x \cos x + 4\ln x$$

In [35]:
from math import sin, cos, log

x = 1.2
print(sin(x)*cos(x) + 4*log(x))   # log is ln (base e)

1.0670178174513938


## Functions

We have already used Python functions above, e.g. `sqrt` from the `math` module. In general, a function is a collection of statements we can execute wherever and whenever we want. For example, consider any of the formulae we implemented above.

Functions can take any number of inputs (called *arguments* or *parameters*) to produce outputs. Functions help to organise programs, make them more understandable, shorter, and easier to extend. Wouldn't it be nice to implement it just once and then be able to use it again any time you need it, rather than having to write out the whole formula again?

For our first example, we will reuse the formula for the position of a ball in a vertical motion, which we have seen earlier.

In [36]:
def ball_height(v0, t, g=9.81):
    """Function to calculate and return height of the ball in vertical motion.

    Parameters
    ----------
    v0 : float
        Initial velocity (units, m/s).
    t : float
        Time at which we want to know the height of the ball (units, seconds).
    g : float, optional
        Acceleration due to gravity (units, m/s^2). By default 9.81 m/s^2.

    Returns
    -------
    float
        Height of the ball in metres.

    """
    height = v0*t - 0.5*g*t**2

    return height

Let us break this example down:
* Function *signature* (header):
    * Functions start with `def` followed by the name we want to give the function (`ball_height` in this case). Just like with variables, function names must be valid identifiers.
    * Following the name, we have parentheses followed by a colon `(...):` containing zero or more function *arguments*.
    * In this case, `v0` and `t` are *positional arguments*, while `g` is known as a *keyword argument* (more about this later).
* Function *body*:
    * The first thing to notice is that the body of the function is indented one level. All code that is indented with respect to `def`-line belongs to a function.
    * Best practice is to include a [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html) to explain to others (or remind our future self) how to use the function. Docstring is defined as a multi-line string literal (enclosed in triple quotes """).
    * The function output is passed back via the `return` statement.

Notice that this just defines the function. Nothing is executed until we actually *call* the function:

In [37]:
# We pass 5 and 0.6 to v0 and t, respectively.
# Since we do not pass a value for g,
# a default value we defined in function's signature is used.
# The value function returns (height) is put in variable h.
h = ball_height(5, 0.6)

print(f"Ball height: {h:g} metres.")

Ball height: 1.2342 metres.


No return value implies that `None` is returned. `None` is a special Python object (singleton) that semantically often represents an ”empty” or undefined value. It is surprisingly useful, and we will use it a lot later.

Functions can also return multiple values. Let us extend the previous example to calculate the ball's velocity as well as its height:

In [38]:
def ball_height_velocity(v0, t, g=9.81):
    """Function to calculate ball's height and its velocity.

    Parameters
    ----------
    v0 : float
        Initial velocity (units, m/s).
    t : float
        Time at which we want to know the height of the ball (units, seconds).
    g : float, optional
        Acceleration due to gravity (units, m/s^2). By default 9.81 m/s^2.

    Returns
    -------
    float
        Height of ball in metres.
    float
        Velocity of ball in m/s.

    """
    height = v0*t - 0.5*g*t**2
    velocity = v0 - g*t

    return height, velocity


# We pass 5 and 0.6 to v0 and t, respectively.
# The first value function returns (height) is put into variable h,
# whereas the second one (velocity) is placed in v - iterable unpacking.
h, v = ball_height_velocity(5, 0.6)

print("Ball height: %g metres." % h)
print("Ball velocity: %g m/s." % v)

Ball height: 1.2342 metres.
Ball velocity: -0.886 m/s.


## Scope: Local and global variables

Variables defined within a function are said to have *local scope*. That is to say that we can only reference them within that function. Consider the example function defined above where we used the *local* variable *height*. You can see that if you try to print the variable height outside the function, you will get an error.

```python
print(height)

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-50-aa6406a13920> in <module>
----> 1 print(height)

NameError: name 'height' is not defined
```

## Keyword arguments and default input values

Besides *positional arguments*, functions can have arguments of the form `argument_name=value` and they are called *keyword arguments*:

In [39]:
def somefunc(arg1, arg2, kwarg1=True, kwarg2=0):
    print(f"arg1: {arg1}, arg2: {arg2}, kwarg1: {kwarg1}, kwarg2: {kwarg2}")


# Note that we have not specified inputs for kwarg1 and kwarg2.
somefunc("Hello", [1, 2])

arg1: Hello, arg2: [1, 2], kwarg1: True, kwarg2: 0


In [40]:
# Note that we replace the default value for kwarg1.
somefunc("Hello", [1, 2], kwarg1="Hi")

arg1: Hello, arg2: [1, 2], kwarg1: Hi, kwarg2: 0


In [41]:
# Note that we replace the default value for kwarg2.
somefunc("Hello", [1, 2], kwarg2="Hi")

arg1: Hello, arg2: [1, 2], kwarg1: True, kwarg2: Hi


In [42]:
# Here, we replace both default values for keyword arguments kwarg1 and kwarg2.
somefunc("Hello", [1, 2], kwarg2="Hi", kwarg1=6)

arg1: Hello, arg2: [1, 2], kwarg1: 6, kwarg2: Hi


If we use `argument_name=value` for all arguments, their sequence in the function call can be in any order.

In [43]:
somefunc(kwarg2="Hello", arg1="Hi", kwarg1=6, arg2=[2])

arg1: Hi, arg2: [2], kwarg1: 6, kwarg2: Hello


## Boolean expressions

An expression with value `True` or `False` is called a boolean expression. Example expressions for what you would write mathematically as
$C = 40$, $C \ne 40$, $C \ge 40$, $C \gt 40$ and $C \lt 40$ are:

```python
C == 40  # Note: the double == checks for equality!
C != 40  # This could also be written as "not C == 4"
C >= 40
C > 40
C < 40
```

Let us now test some boolean expressions:

In [44]:
C = 41

print("C != 40: ", C != 40)
print("C < 40: ", C < 40)
print("C == 41: ", C == 41)

C != 40:  True
C < 40:  False
C == 41:  True


Several conditions can be combined with keywords `and` and `or` into a single boolean expression:

* **Rule 1**: (`C1 and C2`) is `True` only if both `C1` and `C2` are `True`.
* **Rule 2**: (`C1 or C2`) is `True` if either `C1` or `C2` are `True`.

Examples:

In [45]:
x = 0
y = 1.2

print("x >= 0 and y < 1:", x >= 0 and y < 1)
print("x >= 0 or y < 1:", x >= 0 or y < 1)

x >= 0 and y < 1: False
x >= 0 or y < 1: True


## Loops
Suppose we want to make the following table of Celsius and Fahrenheit degrees:
```
 -20  -4.0
 -15   5.0
 -10  14.0
  -5  23.0
   0  32.0
   5  41.0
  10  50.0
  15  59.0
  20  68.0
  25  77.0
  30  86.0
  35  95.0
  40 104.0
```

How do we write a program that prints out such a table? We know that $F = \frac{9}{5}C + 32$, and a single line in this table is:

In [46]:
C = -20
F = 9/5*C + 32

print(C, F)

-20 -4.0


Now, we can just repeat these statements:

In [47]:
C = -20; F = 9/5*C + 32; print(C, F)
C = -15; F = 9/5*C + 32; print(C, F)
C = -10; F = 9/5*C + 32; print(C, F)
C = -5; F = 9/5*C + 32; print(C, F)
C = 0; F = 9/5*C + 32; print(C, F)
C = 5; F = 9/5*C + 32; print(C, F)
C = 10; F = 9/5*C + 32; print(C, F)
C = 15; F = 9/5*C + 32; print(C, F)
C = 20; F = 9/5*C + 32; print(C, F)
C = 25; F = 9/5*C + 32; print(C, F)
C = 30; F = 9/5*C + 32; print(C, F)
C = 35; F = 9/5*C + 32; print(C, F)
C = 40; F = 9/5*C + 32; print(C, F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


We can see that works but it is **very boring** to write and very easy to introduce a misprint.

**You really should not be doing boring repetitive tasks like this.** Spend your time instead looking for a smarter solution. When programming becomes boring, there is usually a construct that automates the writing. Computers are very good at performing repetitive tasks. For this purpose we use **loops**.

## The `while` loop

A `while`-loop executes repeatedly a set of statements as long as a boolean `condition` is `True`

```python
while condition:
    <statement 1>
    <statement 2>
    ...

<first statement after the loop>
```

Note that all statements to be executed within the loop must be indented by the same amount! The loop ends when an unindented statement is encountered.

In Python, indentations are very important. For instance, when writing a `while` loop:

In [48]:
counter = 0
while counter <= 10:
    counter = counter + 1

print(counter)

11


Let us take a look at what happens when we forget to indent correctly:

```python
counter = 0
while counter <= 10:
counter = counter + 1
print(counter)


  File "<ipython-input-86-d8461f52562c>", line 3
    counter = counter + 1
    ^
IndentationError: expected an indented block
```

Let us now use the `while`-loop to create the table above:

In [49]:
C = -20                 # Initialise C
dC = 5                  # Increment for C within the loop
while C <= 40:          # Loop heading with condition (C <= 40)
    F = (9/5)*C + 32  # 1st statement inside loop
    print(C, F)         # 2nd statement inside loop
    C = C + dC          # Increment C for the next iteration of the loop.

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


## Lists
So far, one variable has referred to one number (or string). Sometimes, however, we naturally have a collection of numbers, e.g. degrees −20, −15, −10, −5, 0, ..., 40. One way to store these values in a computer program would be to have one variable per value:

In [50]:
C1 = -20
C2 = -15
C3 = -10
...
C13 = 40

This is clearly a terrible solution, particularly if we have lots of values. A better way of doing this is to collect values together in a list:

In [51]:
C = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]

Now, there is just one variable, `C`, holding all the values. We access elements in a list via an index. List indices are always *zero-indexed*, i.e. they are numbered as 0, 1, 2, and so forth up to the number of elements minus one:

In [52]:
mylist = [4, 6, -3.5]
print("First element:", mylist[0])
print("Second element:", mylist[1])
print("Third element:", mylist[2])

First element: 4
Second element: 6
Third element: -3.5


Here are a few examples of operations that you can perform on lists:

In [53]:
C = [-10, -5, 0, 5, 10, 15, 20, 25, 30]
C.append(35)  # add new element 35 at the end
print(C)

[-10, -5, 0, 5, 10, 15, 20, 25, 30, 35]


In [54]:
C = C + [40, 45]  # And another list to the end of C
print(C)

[-10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [55]:
C.insert(0, -15)  # Insert -15 as index 0
print(C)

[-15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [56]:
del C[2]  # delete 3rd element
print(C)

[-15, -10, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [57]:
del C[2]  # delete what is now 3rd element
print(C)

[-15, -10, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [58]:
print(len(C))  # length of list

11


In [59]:
print(C.index(10))  # Find the index of the element with the value 10

3


In [60]:
print(10 in C)  # True only if the value 10 is stored in the list

True


In [61]:
print(C[-1])  # The last value in the list

45


In [62]:
print(C[-2])  # The second last value in the list

40


We can also extract sublists using `:`:

In [63]:
print(C[5:])  # From index 5 to the end of the list

[20, 25, 30, 35, 40, 45]


In [64]:
print(C[5:7])  # From index 5 up to, but NOT including index 7

[20, 25]


In [65]:
print(C[7:-1])  # From index 7 up to the second last element

[30, 35, 40]


In [66]:
print(C[:])  # [:] specifies the whole list.

[-15, -10, 5, 10, 15, 20, 25, 30, 35, 40, 45]


We can also *unpack* the elements of a list into separate variables:

In [67]:
somelist = ["Curly", "Larry", "Moe"]
stooge1, stooge2, stooge3 = somelist
print(f"stooge1: {stooge1}, stooge2: {stooge2}, stooge3: {stooge3}")

stooge1: Curly, stooge2: Larry, stooge3: Moe


## `for`-loops
We can visit each element in a list and process it with some statements using a `for`-loop, for example:

In [68]:
degrees = [0, 10, 20, 40, 100]
for C in degrees:
    print("list element:", C)
print(f"The list has {len(degrees)} elements.")

list element: 0
list element: 10
list element: 20
list element: 40
list element: 100
The list has 5 elements.


Notice again how the statements to be executed within the loop must be indented! Let us now revisit the conversion table example using the `for` loop:

In [69]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
for C in Cdegrees:
    F = (9/5)*C + 32
    print(C, F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


We can easily beautify the table using the *printf syntax* we encountered previously:

In [70]:
for C in Cdegrees:
    F = (9.0/5)*C + 32
    print(f"{C:5d} {F:5.1f}")

  -20  -4.0
  -15   5.0
  -10  14.0
   -5  23.0
    0  32.0
    5  41.0
   10  50.0
   15  59.0
   20  68.0
   25  77.0
   30  86.0
   35  95.0
   40 104.0


It is also possible to rewrite the `for` loop as a `while` loop, i.e.

```python
for element in somelist:
    # process element
```

can always be transformed to a `while` loop
```python
index = 0
while index < len(somelist):
    element = somelist[index]
    # process element
    index += 1
```

Let us see how a previous example would look like if we used `while` instead of `for` loop:

In [71]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
index = 0
while index < len(Cdegrees):
    C = Cdegrees[index]
    F = (9.0/5)*C + 32
    print(f"{C:5d} {F:5.1f}")
    index += 1

  -20  -4.0
  -15   5.0
  -10  14.0
   -5  23.0
    0  32.0
    5  41.0
   10  50.0
   15  59.0
   20  68.0
   25  77.0
   30  86.0
   35  95.0
   40 104.0


Rather than just printing out the Fahrenheit values, let us also store these computed values in a list of their own:

In [72]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
Fdegrees = []  # start with empty list
for C in Cdegrees:
    F = (9/5)*C + 32
    Fdegrees.append(F)  # add new element to Fdegrees
print(Fdegrees)

[-4.0, 5.0, 14.0, 23.0, 32.0, 41.0, 50.0, 59.0, 68.0, 77.0, 86.0, 95.0, 104.0]


In Python, `for` loops usually loop over list values (elements), i.e.

```python
for element in somelist:
    # process variable element
```

However, we can also loop over list indices:

```python
for i in range(0, len(somelist), 1):
    element = somelist[i]
    # process element or somelist[i] directly
```

The statement `range(start, stop, increment)` generates a list of integers *start*, *start+increment*, *start+2\*increment*, and so on up to, but not including, *stop*. We can also write `range(stop)` as an abbreviation for `range(0, stop, 1)`:

In [73]:
for i in range(3):  # same as range(0, 3, 1)
    print(i)

0
1
2


In [74]:
for i in range(2, 8, 3):
    print(i)

2
5


## List comprehensions
Consider this example where we compute two lists in a `for` loop:

In [75]:
n = 16

# empty lists
Cdegrees = []
Fdegrees = []

for i in range(n):
    Cdegrees.append(-5 + i*0.5)
    Fdegrees.append((9/5)*Cdegrees[i] + 32)

print("Cdegrees = ", Cdegrees)
print("Fdegrees = ", Fdegrees)

Cdegrees =  [-5.0, -4.5, -4.0, -3.5, -3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5]
Fdegrees =  [23.0, 23.9, 24.8, 25.7, 26.6, 27.5, 28.4, 29.3, 30.2, 31.1, 32.0, 32.9, 33.8, 34.7, 35.6, 36.5]


As constructing lists is a very common task, the above way of doing it can become very tedious both to write and read. Therefore, Python has a compact construct, called *list comprehension* for generating lists from a `for` loop:

In [76]:
n = 16
Cdegrees = [-5 + i*0.5 for i in range(n)]
Fdegrees = [(9/5)*C + 32 for C in Cdegrees]
print("Cdegrees = ", Cdegrees)
print("Fdegrees = ", Fdegrees)

Cdegrees =  [-5.0, -4.5, -4.0, -3.5, -3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5]
Fdegrees =  [23.0, 23.9, 24.8, 25.7, 26.6, 27.5, 28.4, 29.3, 30.2, 31.1, 32.0, 32.9, 33.8, 34.7, 35.6, 36.5]


The most general form of a list comprehension is:
```python
somelist = [expression for element in somelist if condition]
```

Here the `condition` can be used to pick out elements which satisfy a specific property.